### PR TITLE
restartable builder, restartable tasks, cache albums, expose tags

### PR DESCRIPTION
**Pull request by bruce-szalwinski**
ChangeLog
======
- Created separate module for generation `google_token.json`
- apiclient changed its name to googleapiclient, so updated code accordingly
- `discovery.build` in the google-api-python-client doesn't allow file_cache, so disabling
- make `build_migration_photo_list` restartable
- build cache of google album titles to reduce api calls, increase speed
- catch/retry http failures in celery tasks using exponential backoff

### CODE REVIEW

File: build_flickr_verifier.py [hunk 1]
```diff
@@ -1,13 +1,13 @@
 import flickr_api
 import os
-flickr_api.set_keys(api_key=os.environ['FLICKR_API_KEY'], api_secret=os.environ['FLICKR_API_SECRET'])
-a = flickr_api.auth.AuthHandler() # creates a new AuthHandler object
-perms = "read" # set the required permissions
-url = a.get_authorization_url(perms)
-print(url)
+api_key=os.environ['FLICKR_API_KEY']
```
Comment (bruce-szalwinski) [inline build_flickr_verifier.py:3]: do the flickr auth all in one script, prompt for verifier code.

File: build_migration_photos_list.py [hunk 2]
```diff
@@ -26,6 +26,9 @@
         except flickr_api.flickrerrors.FlickrServerError as e:
             print(f"Couldn't get original size URL {e}. Skipping")
             continue
+        except flickr_api.flickrerrors.FlickrError as e:
```
Comment (bruce-szalwinski) [inline build_migration_photos_list.py]: if can't find original size, then skip rather than crash.

File: migration_util.py [hunk 3]
```diff
@@ -1,26 +1,23 @@
 from __future__ import print_function
-from apiclient.discovery import build
+from googleapiclient.discovery import build
```
Comment (bruce-szalwinski) [inline migration_util.py:2]: google api client changed names, so updating

File: migration_util.py [hunk 4]
```diff
@@ -1,26 +1,23 @@
...
 from httplib2 import Http
 from oauth2client import file, client, tools
 import requests
 import urllib
 from io import BytesIO
 from pathlib import Path
 
+
 def authorize_with_google():
-    # Setup the Photo v1 API
```
Comment (bruce-szalwinski) [inline migration_util.py:11]: run the `oauth.py` to generate token, here we'll just read it.

File: migration_util.py [hunk 5]
```diff
@@ -1,26 +1,23 @@
...
-        creds = tools.run_flow(flow, store)
+    return store.get()
 
-    return creds
 
 def get_google_photos_service(google_creds):
-    return build('photoslibrary', 'v1', http=google_creds.authorize(Http()))
+    return build('photoslibrary', 'v1',
+                 http=google_creds.authorize(Http()),
+                 cache_discovery=False)
```
Comment (bruce-szalwinski) [inline migration_util.py:22]: no cache discovery, https://github.com/googleapis/google-api-python-client/issues/299, avoid the exceptions / warnings.

File: oauth.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+from __future__ import print_function
```
Comment (bruce-szalwinski) [inline oauth.py:1]: moved from migration_util into its own module.

File: celery_migration_app.py [hunk 7]
```diff
@@ -20,7 +22,7 @@
     'accept_content': ['json']})
 
 
-@app.task
+@app.task(autoretry_for=(RequestException,HttpError), retry_backoff=True)
```
Comment (bruce-szalwinski) [inline celery_migration_app.py]: Tried to run with concurrency=4, ran into quota issues, so implementing exponential backoff.

"Quota exceeded for quota metric 'photoslibrary.googleapis.com/write_requests' and limit 'WritesPerMinutePerUser' of service 'photoslibrary.googleapis.com'"

File: build_migration_photos_list.py [hunk 8]
```diff
@@ -1,45 +1,55 @@
 from __future__ import print_function
```
Comment (bruce-szalwinski) [inline build_migration_photos_list.py:1]: create pickle file for each album (photoset) in flickr.  record which albums have been built to make the process restartable.

File: celery_migration_app.py [hunk 9]
```diff
@@ -19,17 +21,21 @@
     'result_serializer': 'json',
     'accept_content': ['json']})
 
+r = redis.Redis(host='0.0.0.0', port=6379, db=0, decode_responses=True)
 
-@app.task
-def migrate_photo(photo_title, photo_url, album_title):
+
+@app.task(autoretry_for=(RequestException, HttpError),
```
Comment (bruce-szalwinski) [inline celery_migration_app.py:27]: retry tasks for api failures.  Can run with concurrency > 1 and tasks will do exponential backoff if being throttled.

File: build_migration_photos_list.py [hunk 10]
```diff
@@ -1,45 +1,55 @@
...
+                print(f"Couldn't get original size URL {e}. Skipping")
+                continue
+
+            my_photos.append({
+                "album": photo_set['title'],
+                "photoId": photo['id'],
+                "photoTitle": photo['title'],
+                "photoUrl": photo_url,
+                "processed": False,
+                "photoTags": ','.join([t['text'] for t in photo.getTags()]),
```
Comment (bruce-szalwinski) [inline build_migration_photos_list.py:43]: pass tags in Flickr to Google.

File: celery_migration_app.py [hunk 11]
```diff
@@ -19,17 +21,21 @@
...
+          retry_kwargs={'max_retries': 20})
+def migrate_photo(photo_title, photo_url, album_title, photo_tags):
     google_creds = authorize_with_google()
     service = get_google_photos_service(google_creds)
 
-    album_id = find_album_on_google(service, album_title)
-
-    if album_id is None:
-        album_id = create_album_on_google(service, album_title)
+    with r.lock("find-album"):
```
Comment (bruce-szalwinski) [inline celery_migration_app.py:34]: attempt to find the album in the local album cache, if not found then make the trip to google and store the album title / id in redis.  saves a lot of trips to google api.

File: create_album_cache.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
+import redis
```
Comment (bruce-szalwinski) [inline create_album_cache.py:1]: rummage through google and create a cache of album titles and ids.

File: create_migration_tasks.py [hunk 13]
```diff
@@ -1,10 +1,18 @@
 from celery_migration_app import migrate_photo
 import pickle
-from pathlib import Path
+import os
 
-with open((Path.cwd() / "photos_to_move.pickle").resolve(), "rb") as photo_tasks_file:
-    my_photos = pickle.load(photo_tasks_file)
+for file in os.listdir("photosets/"):
```
Comment (bruce-szalwinski) [inline create_migration_tasks.py:5]: rummage through the photosets/*.pickle files and create migration tasks.

File: migration_util.py [hunk 14]
```diff
@@ -56,27 +46,29 @@ def upload_photo_to_google(google_auth, service, album_id, photo_data, photo_tit
...
-        payload = {"albumId": album_id,
-                   "newMediaItems": [
-                       {"description": "test",
-                        "simpleMediaItem": {"uploadToken": upload_token}
-                        }
-                   ]}
+        payload = {
+            "albumId": album_id,
+            "newMediaItems": [{
+                "description": photo_tags,
```
Comment (bruce-szalwinski) [inline migration_util.py:56]: description is now tags or title from flickr.

File: move_failed_tasks_back.py [hunk 15]
```diff
@@ -26,3 +39,11 @@
 
 for task_file in failed_task_files:
     move(task_file.as_posix(), (Path.cwd() / "celery" / "out").as_posix())
+
+# remove errant messages
```
Comment (bruce-szalwinski) [inline move_failed_tasks_back.py:43]: remove the message that we are attempting to retry so that we don't retry it again.

File: task_checker.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
+import pickle
```
Comment (bruce-szalwinski) [inline task_checker.py:1]: purpose is to check to see that at least one photo from a photo set has been migrated.  It does this by creating a list of URLs from all of the completed tasks.  Then for each photoset, get the url of the first photo and see if it exists as a completed task.

**Comment by yakneens** [comment]
Great contributions @bruce-szalwinski. My attention is away from this project as I have my PhD defence in a couple of weeks, but I appreciate you pushing through despite initial setbacks, and making all these improvements.

**Comment by bruce-szalwinski** [comment]
> Great contributions @bruce-szalwinski. My attention is away from this project as I have my PhD defence in a couple of weeks, but I appreciate you pushing through despite initial setbacks, and making all these improvements.

I enjoyed reading your dissertation.  Best of luck with your project!

**Pull request merged**